File: Data Structures & Algorithms/binary-tree-diameter/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int bestDia = 0;
    
    int diameterOfBinaryTree(TreeNode* root) {
        depth(root);
        return bestDia;
    }

    int depth(TreeNode* root){
        if(root == nullptr){
            return 0;
        }

        int leftDepth = depth(root->left);
        int rightDepth = depth(root->right);

        int currDia = leftDepth + rightDepth;
        if(currDia > bestDia){
            bestDia = currDia;
        }
        return max(depth(root->left), depth(root->right))+1;
    }
};

File: Data Structures & Algorithms/search-2d-matrix/submission-2.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        n = len(matrix)
        m = len(matrix[0])

        left, right = 0, m*n - 1

        while left <= right:
            mid = (left + right) // 2
            val = matrix[mid//m][mid % m]

            if val == target:
                return True
            elif val < target:
                left = mid + 1
            elif val > target:
                right = mid - 1

        return False



        